enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,266 @@
+---
+title: enable-the-ability-to-configure-dns-over-tls-for-forwarded-requests
+authors:
+  - "@brandisher"
+reviewers:
+  - "@Miciah"
+  - "@rfredette"
+  - "@candita"
+  - "@frobware"
+  - "@knobunc"
+approvers:
+  - "@Miciah"
+  - "@knobunc"
+creation-date: 2021-12-08
+last-updated: 2021-12-08
+tracking-link:
+  - https://issues.redhat.com/browse/NE-750
+see-also:
+  - "/enhancements/dns/plugins.md"
+---
+
+# Enable the ability to configure DNS-over-TLS for forwarded requests
+
+## Summary
+
+This enhancement adds optional configuration to enable DNS-over-TLS when 
+either `spec.servers[].forwardPlugin` and/or `spec.upstreamResolvers` are 
+configured in the cluster DNS spec.
+
+## Motivation
+
+In environments that are highly regulated or geographically dispersed there
+may be a need to forward requests to upstream DNS resolvers. Because DNS is
+unencrypted by default there are concerns in these environments around data
+privacy and integrity. Currently, there is no way to configure DNS-over-TLS for
+upstream resolvers so there is also no workaround to alleviate this security
+concern. The expansion of the `UpstreamResolver` and `ForwardPlugin` APIs  for 
+DNS request forwarding improves the current security posture for DNS on 
+OpenShift.
+
+### Goals
+
+1. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream FQDN for the default zone in the DNS configuration.
+2. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream FQDN for a non-default zone.
+3. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for each listed upstream in the default zone.
+4. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for a set of upstreams on a non-default zone.
+
+### Non-Goals
+
+* Allowing configuration of DNS-over-TLS for use cases other than OpenShift's default DNS provider to upstream resolvers is out of scope.
+
+## Proposal
+
+This proposal covers changes to two APIs. The first is a change to 
+`UpstreamResolvers`.  To enable a cluster administrator to configure 
+DNS-over-TLS, the `UpstreamResolvers` type will get three new optional members:
+
+`TLSSecurityProfile` encapsulates TLS specific configuration. Including this 
+allows customers to set their desired TLS version, ciphers, etc.
+
+`CABundle` references a ConfigMap that contains a single CA certificate 
+or a CA bundle.
+
+`ServerName` contains an FQDN to validate when establishing the connection. 
+This is relevant in particular with cases where a customer has multiple 
+upstreams behind one serving certificate (e.g. `1.1.1.1` and `2.2.2.2` 
+behind `dns.myorg.com`)
```

Comment:
`ServerName` serves two purposes:
* The client specifies the server name in the TLS handshake, which tells the remote end which server the client wants.  (I don't know whether virtual servers are common with DNS, but theoretically the remote end might need this.)
* The client uses the server name to verify the serving certificate that the server provides.